### PR TITLE
Check raw pointers coming from ffi in get_reply.

### LINKED ISSUE #64: NULL pointer dereference when XCB connection is in an error state
The following program shows two and a half ways to get `NULL` pointer dereferences in safe rust code via rust-xcb. They are all about somehow getting the underlying `xcb_connection_t` into an error state, after which `xcb_*_reply` will start returning both reply and error as `NULL`. The code in rust-xcb assumes that when no error was returned, the reply is a valid pointer.
```rust
extern crate xcb;

use std::iter::repeat;

fn main() {
    let (conn, screen_no) = xcb::Connection::connect(None).unwrap();
    let setup = conn.get_setup();
    let screen = setup.roots().nth(screen_no as usize).unwrap();

    let variant = 1; // Choose a variant of your liking here
    match variant {
        1 => {
            // My X11 server does not support XEVIE. Use this to break the connection.
            xcb::xevie::query_version(&conn, 0, 0);
            println!("Connection should be broken now: {:?}", conn.has_error());
            // With the broken connection, xcb_get_input_focus() will return NULL and also error=NULL
            println!("{:?}", xcb::get_input_focus(&conn).get_reply().unwrap().focus());
        },
        2 => {
            // Accessing the reply to XEVIE's QueryVersion already segfaults (same kind of NULL pointer)
            let query_version = xcb::xevie::query_version(&conn, 0, 0);
            println!("Connection should be broken now: {:?}", conn.has_error());
            println!("{:?}", query_version.get_reply().unwrap().server_major_version());
        },
        3 => {
            // Exceed the maximum request size (just to show that there are more ways to break the
            // connection than "unsupported extension"; it could also just be the X11 server dying)
            let max_length = conn.get_maximum_request_length() as usize;
            let larger_size = (max_length + 1) * 4;
            let data: Vec<_> = repeat(0).take(larger_size).collect();
            xcb::put_image(&conn, 0, screen.root(), 0, 0, 0, 0, 0, 0, 0, &data);
            println!("Connection should be broken now: {:?}", conn.has_error());
            println!("{:?}", xcb::get_input_focus(&conn).get_reply().unwrap().focus());
        },
        _ => panic!()
    }
}
```

### PR DESCRIPTION
**Pull request by Lompik**
```xcb``` will always return ```NULL``` if connection has errors... so also check for errors in ```VoidCookie::request_check```

Fixes #64

### CODE REVIEW

File: rs_client.py [hunk 1]
```diff
@@ -1963,6 +1963,7 @@ def _cookie(request):
                 _r("    let reply = %s {", reply)
                 _r("        ptr: %s (self.conn.get_raw_conn(), self.cookie, &mut err)", func)
                 _r("    };")
+                _r("    if reply.ptr.is_null() { return Err(base::GenericError { ptr: err }) }")
```
Comment (psychon) [inline rs_client.py]: Uhm... This hands out a `GenericError` with a `NULL` pointer. Wouldn't that just cause crashes when this one is used?

**Comment by psychon** [comment]
Your patch always returns `Err(ReplyError::NullResponse)` when `reply.ptr.is_null()`. The case with the non-`NULL` error is unreachable

Might I suggest the following (untested and likely does not compile, but this is about the idea):
```rust
match (reply.ptr.is_null(), err.is_null) {
  (false, false) => unreachable!(), // xcb does not do this
  (true, true) => Err(base::ReplyError::NullResponse),
  (true, false) => Err(base::ReplyError::GenericError(base::GenericError { ptr: err })),
  (false, true) => Ok(wrap-the-pointer-in-the-right-struct)
}
```
Also, this does not fix "the idea behind" #64: This only touches the code for checked requests. Unchecked requests are still assumed to always produce a reply. I expect one can easily update the code in #64 to also trigger this problem.

**Comment by meh** [comment]
Any update on this? :panda_face:

**Comment by rtbo** [comment]
Thank you for your pull request.
I'm sorry for the long reply time. I have difficulties to give time to this project

File: src/base.rs [hunk 2]
```diff
@@ -200,8 +200,9 @@ impl<'a> VoidCookie<'a> {
             let c: xcb_void_cookie_t = mem::transmute(self.cookie);
             let err = xcb_request_check(self.conn.get_raw_conn(), c);
 
+            let conn_has_error = self.conn.has_error().is_ok();
             std::mem::forget(self);
-            if err.is_null() {
+            if err.is_null() && conn_has_error {
```
[the remark this answers is not in the document]
Comment (Lompik) [reply src/base.rs]: Good catch. 
This variable was misnamed and hold true if the connection has no error.
XCB does return a null pointer if the connection has errors. I made some changes to handle that case.

**Comment by rtbo** [comment]
Looks good, thank you

**Pull request merged**